File: Rahul/StrategyPattern/Duck.java
package strategy.pattern;

public class Duck {
	IFlyBehaviour flyBehaviour;
	IQuackBehaviour quackBehaviour;
	
	public Duck(IFlyBehaviour flyBehaviour,
			IQuackBehaviour quackBehaviour) {
		this.flyBehaviour = flyBehaviour;
		this.quackBehaviour = quackBehaviour;
	}
	public void setFlyBehaviour(IFlyBehaviour flyBehaviour) {
		this.flyBehaviour = flyBehaviour;
	}
	
	public void setQuackBehaviour(IQuackBehaviour quackBehaviour) {
		this.quackBehaviour = quackBehaviour;
	}
	
	public void fly() {
		this.flyBehaviour.fly();
	}
	public void quack() {
		this.quackBehaviour.quack();
	}
	
	public void swim() {
		System.out.println("Swimming");
	}
}

File: Rahul/StrategyPatternCPP/Duck.cpp
#ifndef __DUCK__
#define __DUCK__
#include "IFlyBehaviour.cpp"
#include "IQuackBehaviour.cpp"
#include <iostream>

class Duck{
    private:
    IFlyBehaviour *flyBehaviour;
    IQuackBehaviour *quackBehaviour;
    public:
    Duck(IFlyBehaviour *flyBehaviour , IQuackBehaviour *quackBehaviour):
    flyBehaviour(flyBehaviour) , quackBehaviour(quackBehaviour){
    }
    void setFlyBehaviour(IFlyBehaviour *flyBehaviour){
        delete( this->flyBehaviour );
        this->flyBehaviour = flyBehaviour;
    }
    void setQuackBehaviour(IQuackBehaviour *quackBehaviour){
        delete( this->quackBehaviour );
        this->quackBehaviour = quackBehaviour;
    }
    void fly(){
        this->flyBehaviour->fly();
    }
    void quack(){
        this->quackBehaviour->quack();
    }
    void swim(){
        std::cout<<"Swim\n";
    }
};
#endif